feat: add support for auto-closing brackets/quotations in the REPL

Snehil-Shah opened the pull request:
Resolves #1672 

## Description

> What is the purpose of this pull request?

This pull request:

-   Implements auto-closing of brackets and quotations in the REPL.
    
![demo](https://github.com/stdlib-js/stdlib/assets/130062020/8df01d0b-1a7a-496f-8a34-bc8890c52cfc)

-   Adds option to toggle this behavior.

![settings](https://github.com/stdlib-js/stdlib/assets/130062020/090312c0-f89c-47d1-8a8a-8f5af2152708)

-   Adds support for modifier key to add newline for multi-line editing

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1672
-   relates to https://github.com/stdlib-js/google-summer-of-code/issues/1

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

For **Multi-line editing**, I have used **ALT+ENTER** as the modifier key.

This might seem odd given SHIFT+ENTER is the conventional key combination to add a newline instead of executing a command, but it is done because the terminal or the readline module cannot differentiate between ENTER & SHIFT+ENTER or even CTRL+ENTER.

![values of keypress events](https://github.com/stdlib-js/stdlib/assets/130062020/1d12f108-d399-40ee-80ee-6d7541c1f3c2)

ALT+ENTER generates a different keypress value than the others.

## Questions

> Any questions for reviewers of this pull request?

I am utilizing readline module's private methods (shown in image) to mess with the input stream. Public apis like `write()` or `cursorTo()` aren't working as expected.

![image](https://github.com/stdlib-js/stdlib/assets/130062020/cacca31a-e35a-4ea2-8fed-0f20ee32e772)

I am not entirely sure if it's okay to use these private methods, but they definitely work as desired.

**UPDATE**: Solved!

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Snehil-Shah committed:
feat: add auto-closing brackets & quotations in REPL

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
feat: add option to toggle auto-closing behavior

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented:
@kgryte maybe we can also add a keyboard shortcut to toggle this behaviour?

kgryte commented:
@Snehil-Shah Atm, we don't have any support for keyboard shortcuts in the REPL, so I'd like to avoid adding anything like that here. Adding keyboard shortcuts is a future concern, but needs to be addressed more holistically (e.g., `vim`, Sublime  Text, etc, keybindings).

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -268,6 +271,35 @@ function REPL( options ) {
+
+		// For ALT+ENTER keypress, add a new line in multi-line mode:
+		if ( key && key.name === 'return' && key.meta ) {
+			self._rli._insertString( '\n' );
```
As you allude to in the OP, we should not be using private methods on the readline instance.

kgryte replied:
My guess is that it does not work with `write`, as writing `\n` may trigger the `line` event.

kgryte replied:
For reference as to how we currently handle multi-line input, see https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/repl/lib/process_line.js and https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/repl/lib/display_prompt.js.

To write a newline, you can write to the `ostream`. I imagine the same could be done for bracket matching.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands.js:
```diff
@@ -122,6 +123,7 @@ function commands( repl ) {
 	cmds.push( [ 'rerequire', onRerequire( repl ), false ] );
 	cmds.push( [ 'rerun', onRerun( repl ), false ] );
 	cmds.push( [ 'reset', onReset( repl ), false ] );
+	cmds.push( [ 'toggleAutoClose', onToggleAutoClose( repl ), false ] );
```
Ideally, we would not have separate commands for individual settings. Instead, we should have a single command (e.g., getter: `settings( name )`, setter: `settings( name, value )`) which can handle this and any future setting. Otherwise, we'll just clutter the REPL namespace.

Snehil-Shah replied:
![image](https://github.com/stdlib-js/stdlib/assets/130062020/948cc53d-d7f9-4038-a422-39593f53ed32)

@kgryte something like this is okay? If yes I will update the README..

kgryte replied:
@Snehil-Shah This seems reasonable. What is this setting named in other contexts? E.g., in Sublime Text, this is named `auto_match_enabled`. I might rename `autoCloseSymbols` to `autoMatch` and update the description to "Automatically insert matching brackets, parentheses, and quotes."

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -158,6 +159,7 @@ function REPL( options ) {
 	setNonEnumerableReadOnly( this, '_isTTY', opts.isTTY );
 	setNonEnumerableReadOnly( this, '_sandbox', opts.sandbox );
 	setNonEnumerable( this, '_quiet', opts.quiet ); // allow this to be internally toggled
+	setNonEnumerable( this, '_autoClose', opts.autoClose );
```
We should have a private `settings` object for storing these kinds of option values.

kgryte reviewed (changes requested):
Left some initial comments.

Snehil-Shah committed:
refactor: prevent use of private readline methods

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
feat: add generic `settings` command to manage all settings

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
fix: remove autoClose from defaults

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
fix: remove `autoClose` from REPL opts

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+// VARIABLES //
+
+var debug = logger( 'repl:command:settings' );
+var SETTINGS = {
+	'autoCloseSymbols': {
+		'desc': 'Auto-close brackets and quotations.'
+	}
+};
```
@kgryte right now I have hard coded the settings list, maybe we can list them in a json file?

kgryte replied:
This is fine for now. I'd also add the default settings value.

```javascript
var SETTINGS = {
	'autoCloseSymbols': {
		'desc': 'Auto-close brackets and quotations.',
		'default': true
	}
};
```

kgryte replied:
We should also include a "type" and then we can do settings-specific validation below, instead of assuming its a boolean.

Snehil-Shah committed:
fix: error messages

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+			return;
+		}
+		repl._settings[name] = value;
+		repl._ostream.write( format( '\nUpdated setting %s to %s.\n\n', name, ( value ) ? 'true' : 'false' ) );
```
@kgryte Btw, `format` doesn't automatically convert boolean to string, and afaik there isn't a boolean specifier in the format method. is this expected behaviour? If not, maybe it can be related to #845..

kgryte replied:
@Snehil-Shah Demonstrating string serialization of booleans via `@stdlib/string/format`:

```
In [14]: var fmt = require( '@stdlib/string/format' )
Out[14]: [Function: format]

In [15]: fmt( '%s', true )
Out[15]: 'true'
```

Seems to work to me.

Snehil-Shah replied:
> @Snehil-Shah Demonstrating string serialization of booleans via `@stdlib/string/format`:
> 
> ```
> In [14]: var fmt = require( '@stdlib/string/format' )
> Out[14]: [Function: format]
> 
> In [15]: fmt( '%s', true )
> Out[15]: 'true'
> ```
> 
> Seems to work to me.

Oh.. will check👍

Snehil-Shah replied:
@kgryte I tried it when assigning it to a variable, it doesn't work. is it a bug or I am messing up somewhere?

![image](https://github.com/stdlib-js/stdlib/assets/130062020/15d8ffc0-177c-4ef2-80f1-0a46c854f1f5)

kgryte replied:
One small adjustment:

```
In [5]: var fmt = require( '@stdlib/string/format' );

In [6]: var a = true;

In [7]: fmt( '%s', a )
Out[7]: 'true'
```

Notice the missing semicolon. Semicolons silence printed output.

Snehil-Shah replied:
Oh! got it, thank you so much

Snehil-Shah commented:
@kgryte IPython uses CTRL+O to manually add multi-line, maybe we can add that too on top of ALT+ENTER?

kgryte commented:
Hmm...CTRL+O is not obvious to me. I think people are more likely to be used to chat apps which use CMD or SHIFT as the modifier with ENTER.

Snehil-Shah commented:
@kgryte I have another improvement in mind. 

when the user is typing, there is a good chance they might instinctively add the closing bracket too after the opening bracket, when say calling a function with no arguments.
with the current implementation, they end up adding an extra closing symbol.

I propose improving this behavior similar to IDEs where if you type the opening and closing symbol with flow, the symbol added automatically is overridden.

Snehil-Shah committed:
feat: manual multiline mode from cursor position

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented:
@kgryte improved the manual multi-line editing by making it cursor-aware.

![manual](https://github.com/stdlib-js/stdlib/assets/130062020/cae0ac26-7093-4bfb-89e4-2c8bef7e43af)

when manually adding a newline (using alt+enter), it adds a newline from the cursor position just like a text editor would!

is this okay?

Snehil-Shah committed:
remove unused flag

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
feat: handle instinctive closing symbols for improved ergonomics

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented:
@kgryte Made auto-closing symbols more ergonomic by handling the case where the user instinctively adds a closing bracket when typing with flow. The extra closing symbol is discarded similar to IDEs.

![instinct](https://github.com/stdlib-js/stdlib/assets/130062020/5cc11bac-669d-4162-a03f-5d0752a31c28)

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -268,6 +280,72 @@ function REPL( options ) {
+			});
+		}
+
+		// For a bracket/quotation input, append corresponding closing symbol
+		if ( self._settings.autoCloseSymbols && symbols[chunk] ) {
+			self._rli.write( symbols[chunk] );
+			self._rli.write( null, {
+				'ctrl': true,
+				'name': 'b'
+			} );
+			self._auto_close_status = true;
+			return;
+		}
+
+		// Handle instinctive closing symbols
+		if ( self._auto_close_status ) {
+			if (
+				typeof chunk !== 'undefined' &&
+				symbols[ self._rli.line[ self._rli.cursor - 2 ] ] === chunk
+			) {
+				self._rli.write( null, {
+					'name': 'backspace'
+				} );
+				self._rli.write( null, {
+					'name': 'right'
+				} );
+			}
+			self._auto_close_status = false;
+		}
+	}
```
@kgryte should I refactor and move this logic out to a separate file?

kgryte replied:
I'd move to a separate function (and file) and make a factory function, where the parent function takes in a REPL instance, creating a closure, and returns this event listener.

Snehil-Shah replied:
@kgryte Should I make a common file for handling all keypress events or just for auto closing symbols?

kgryte replied:
You can start off with a common file for handling all keypress events, although we may need to be cognizant of https://github.com/stdlib-js/stdlib/pull/1832. In which case, refactoring to a common file could be done in a separate follow-up PR.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+	return onCommand;
+
+	/**
+	 * Manages REPL settings.
```
This and the following lines should be aligned with the `/`, not the first `*`.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+	 * If two arguments are given, it updates the value of the given setting name to the given value.
+	 *
+	 * @private
+	 * @param {string} [name] - Setting name
+	 * @param {boolean} [value] - New setting value
+	 * @throws {TypeError} First argument must be a string
+	 * @throws {Error} Must recognize setting name
+	 * @throws {TypeError} Second argument must be a boolean
```
```suggestion
	 * @param {string} [name] - setting name
	 * @param {*} [value] - new setting value
```

Other settings need not have boolean values. E.g., they could be numeric or strings, etc. And we also don't throw an exception in this function; we print an exception message.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+* @private
+* @returns {string} settings help text
+*/
+function help() {
```
I wonder if the printed help text should also display the current settings value for each setting. E.g.,

```text
autoCloseSymbols
	Automatically insert matching brackets, parentheses, and quotes. Value: false.
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+		if ( !isString( name ) ) {
+			err = new TypeError( format( 'invalid argument. First argument must be a string. Value: `%s`.', name ) );
+			debug( 'Error: %s', err.message );
+			repl._ostream.write( format( 'Error: %s\n.', err.message ) );
```
```suggestion
			repl._ostream.write( format( 'Error: %s\n', err.message ) );
```

You don't need trailing periods here or below.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+			repl._ostream.write( format( 'Error: %s\n.', err.message ) );
+			return;
+		}
+		if (arguments.length === 1) {
+			repl._ostream.write('\n'+repl._settings[name]+'\n\n');
```
```suggestion
		if ( arguments.length === 1 ) {
			repl._ostream.write( '\n'+repl._settings[name]+'\n\n' );
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+			repl._ostream.write('\n'+repl._settings[name]+'\n\n');
+			return;
+		}
+		if ( !isBoolean(value) ) {
+			err = new TypeError( format( 'invalid argument. Second argument must be a boolean. Value: `%s`.', name ) );
+			debug( 'Error: %s', err.message );
+			repl._ostream.write( format( 'Error: %s\n.', err.message ) );
+			return;
+		}
+		repl._settings[name] = value;
```
```suggestion
		repl._settings[ name ] = value;
```

kgryte replied:
You cannot assume that all settings will be booleans. As mentioned above, if we include a "type" field in the `SETTINGS` object, we can look-up a validation function from an object based on the value of the type field, thus allowing for settings validation.

kgryte commented on lib/node_modules/@stdlib/repl/lib/commands/settings.js:
```diff
@@ -0,0 +1,136 @@
+			return;
+		}
+		repl._settings[name] = value;
+		repl._ostream.write( format( '\nUpdated setting %s to %s.\n\n', name, ( value ) ? 'true' : 'false' ) );
```
I don't think you need the ternary here (a) and (b) again, we should not assume that all settings will be booleans.

kgryte commented on lib/node_modules/@stdlib/repl/lib/help_text.js:
```diff
@@ -71,6 +71,10 @@ var MSG = [
 	'    userDoc()               Add user-defined documentation.',
 	'    clearUserDocs()         Clear user-defined documentation.',
 	'',
+	'    settings()              List REPL settings.',
+	'    settings(name)          Get setting\'s value.',
+	'    settings(name,value)    Update setting with the new value.',
```
```suggestion
	'    settings()              View and update settings.',
```

Nowhere in this help text do we split things out based on argument signatures, even though there are other variadic functions.

Snehil-Shah replied:
@kgryte I am missing something but `help` , `tutorial`, etc also have settings listed based on arguments right?

kgryte replied:
Ah, you're right. I didn't expand the file to the top. I'd say to just document `settings()` generically and not list the various argument signatures. `help`, `tutorial`, `example` and `info` are effectively privileged as they are expected to be common commands where having call signatures broken out in more detail is warranted. I would not expect that `settings` will be common and we don't privilege the other commands listed with additional documentation for similar reasons.

If a user wants more detailed help text, we should provide such documentation when a user runs `help( settings )`.

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -202,8 +203,18 @@ function REPL( options ) {
-	setNonEnumerable( this, '_multiline_mode', false );
+	// Initialize an internal object storing user settings:
+	setNonEnumerable( this, '_settings', {} );
+	setNonEnumerable( this._settings, 'autoCloseSymbols', true );
```
I'd rename this setting to avoid confusion with built-in `Symbol` instances.

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -202,8 +203,18 @@ function REPL( options ) {
+	setNonEnumerable( this._multiline_mode, 'manual', false );
+	setNonEnumerable( this._multiline_mode, 'buffer', '' );
+
+	// Initialize an internal flag indicating whether auto closing symbols is in progress
```
```suggestion
	// Initialize an internal flag indicating whether auto closing symbols is in progress:
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/process_line.js:
```diff
@@ -62,9 +63,21 @@ function processLine( repl, line ) {
+		repl._multiline_mode.manual = false;
+		displayPrompt( repl, false );
+		repl._ostream.write( repl._multiline_mode.buffer );
+		repl._ostream.write(repeat('\x1b[D', repl._multiline_mode.buffer.length));
```
```suggestion
		repl._ostream.write( repeat( '\x1b[D', repl._multiline_mode.buffer.length ) );
```

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -202,8 +203,18 @@ function REPL( options ) {
+	// Initialize an internal status object for controlling multi-line mode:
+	setNonEnumerable( this, '_multiline_mode', {} );
+	setNonEnumerable( this._multiline_mode, 'status', false );
+	setNonEnumerable( this._multiline_mode, 'manual', false );
```
The use of the name "manual" is not intuitive to me, as it seems to distinguish between when using a modifier key or not. Perhaps rename to `mode` and make the value a string, such as 

```javascript
this._multiline_mode.mode = 'modifer';
```

and for the current behavior of entering multi-line mode for incomplete expressions

```javascript
this._multiline_mode.mode = 'incomplete_expression';
```

or something similar. What I am getting at is that it would be nice to have something more self-documenting to distinguish between the handling.

kgryte reviewed (changes requested):
@Snehil-Shah This is coming along. I agree that the proposed behavior is superior to the status quo, and I agree that adding a settings API is a good addition. I've left some comments which will hopefully help sharpen the implementation.

kgryte commented:
> improved the manual multi-line editing by making it cursor-aware.

@Snehil-Shah I agree that this is preferable.

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -202,8 +203,18 @@ function REPL( options ) {
+	setNonEnumerable( this._settings, 'autoCloseSymbols', true );
+
+	// Initialize an internal status object for controlling multi-line mode:
+	setNonEnumerable( this, '_multiline_mode', {} );
```
Maybe we can also shorten then is to `_multiline`, rather than `multiline_mode`, now that it is an object with various properties.

kgryte commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -202,8 +203,18 @@ function REPL( options ) {
+
+	// Initialize an internal status object for controlling multi-line mode:
+	setNonEnumerable( this, '_multiline_mode', {} );
+	setNonEnumerable( this._multiline_mode, 'status', false );
```
```suggestion
	setNonEnumerable( this._multiline_mode, 'active', false );
```

I wouldn't expect a "status" property or variable to have a boolean value a priori, so I'd change to something which conveys better the value type. Could also do something like `isEnabled` or `isActive`.

Snehil-Shah commented:
> @Snehil-Shah This is coming along. I agree that the proposed behavior is superior to the status quo, and I agree that adding a settings API is a good addition. I've left some comments which will hopefully help sharpen the implementation.

Thanks for the review. On it👍

Snehil-Shah committed:
refactor: suggestions

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
feat: add auto indentation

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah committed:
feat: handle special cases to improve ergonomics

Signed-off-by: Snehil Shah <130062020+Snehil-Shah@users.noreply.github.com>

Snehil-Shah commented:
@kgryte I apologize for the delay.

Did the changes as per suggestions.

Added some new improvements:

- **Auto-indentation**: Automatically indent (4 spaces) when encountering an open bracket to beautify definitions of functions, objects and arrays.

![auto-indent](https://github.com/stdlib-js/stdlib/assets/130062020/76d3650b-25bc-4eff-ad50-abb95f3ce966)

- **Auto-delete paired symbols**: As you are typing and add an extra bracket/quote by mistake, just one backspace removes both the original bracket and the appended one just like an IDE. 

![auto-delete](https://github.com/stdlib-js/stdlib/assets/130062020/4ac47ead-7067-4bbb-abea-cd09011e4beb)

- **Disable auto-closing quotes around possible strings**: sometimes when editing we forget to put strings inside quotes, when adding them later, auto-closing quotes become counter-intuitive, hence is better disabled

![around-text](https://github.com/stdlib-js/stdlib/assets/130062020/30a4d1c4-02b4-4e0e-bf34-34c718284ac6)

**Questions**:

I just have one doubt regarding the refactoring, your previous review mentioned creating factory functions for the keypress listener. I made two files and factory functions (for auto_matching_symbols and manual_mulitline) that returns an onKeypress event listener.
Looking at existing implementations of line, command and close event listeners, all are implemented in the `main.js` file itself, whereas the event handlers (not necessarily factory functions) are implemented in separate files (like `process_line.js` partially handles the line event). I took the same approach..

Snehil-Shah commented:
@kgryte should I also write tests for this?

kgryte committed:
Merge branch 'develop' of https://github.com/stdlib-js/stdlib into snehil-shah-repl

kgryte commented:
@Snehil-Shah I pulled this down and had a play. A few comments:

1. Bracket matching seems broken. When I enter `[`, `{`, or a backtick, it adds two closing brackets, rather than one. It works for a single quote.
2. Deletion seems broken. If I enter a double quote and then press backspace, it only deletes the first quote.
3. The implementation is not syntax aware. Because it works based on analyzing raw text rather than the AST, it attempts auto-closing within string literals, which is not necessarily desired and, at least in my editor, not something which is universally supported elsewhere. Typically, if I am writing plain text (e.g., an e-mail), I am not wanting auto-closing. Similarly, if I am defining a regexp, I often don't want auto-closing because having only one-side of the bracket/quote is what I want, not both.
4. Handling the case where a user types a closing brace/quote seems broken. When I type `'foo'`, including the closing single quote, it creates another single quote rather than ignoring the keypress.
5. Multi-line editing auto-indentation is not syntax aware. E.g., if I enter `function() {\n//...\n}`, that closing brace is also auto-indented, rather than matching the indentation of `function`, similar to how an IDE would automatically dedent.

kgryte commented:
@Snehil-Shah Disregard my previous comment. A bug was introduced during the merge. I'll circle back.

kgryte committed:
fix: consolidate keypress listeners

kgryte committed:
refactor: always attach a keypress listener

kgryte committed:
refactor: document `settings` command signatures

kgryte committed:
refactor: document defaults and add settings object

kgryte committed:
refactor: add module exposing settings validators

kgryte committed:
refactor: add module to validate a settings object

kgryte committed:
fix: update error messages

kgryte committed:
fix: add missing settings file

kgryte committed:
docs: update comment

kgryte committed:
refactor: add support for validating a settings object

kgryte committed:
refactor: remove duplicate option setting and document settings

kgryte committed:
refactor: make the default dependent on whether TTY

kgryte committed:
style: disable lint rules

kgryte committed:
docs: document settings

kgryte committed:
refactor: add module exporting a list of settings names

kgryte committed:
refactor: use module to get list of settings names

kgryte committed:
refactor: update help text generation

kgryte committed:
refactor: only pass down settings object

kgryte committed:
fix: use setting name when constructing error message

kgryte committed:
refactor: add module to test if a value is a setting name

kgryte committed:
refactor: update error messages

kgryte committed:
feat: add `settings` method for getting and setting REPL settings

kgryte committed:
refactor: delegate to `settings` method to validate arguments

kgryte committed:
docs: document `settings` method

kgryte committed:
refactor: handle returning setting value as would normal command

kgryte committed:
refactor: update confirmation message

kgryte committed:
docs: update description

kgryte committed:
refactor: remove support for manually entering multi-line editing

Modifiers were generally not recognized on my Macbook Pro. In general,
we should rethink multi-line editing and, once we do some more R&D,
add support in a follow-up PR.

Ref: https://github.com/stdlib-js/stdlib/issues/2060

kgryte committed:
docs: update copy

Snehil-Shah commented:
@kgryte was it not recognising option+enter on mac?

kgryte committed:
refactor: bind variables to REPL instance to allow "toggabilty"

kgryte committed:
refactor: move symbols table to separate file

kgryte commented:
@Snehil-Shah No, and I tried debugging. Essentially, I couldn't get any modifier to work with ENTER.

kgryte committed:
fix: update require path

Snehil-Shah commented:
> @Snehil-Shah No, and I tried debugging. Essentially, I couldn't get any modifier to work with ENTER.

I assume that's why IPython uses CTRL+O to add new lines instead of any modifier with ENTER, maybe we could do that?

kgryte committed:
refactor: convert auto-matcher to a class

This avoids unnecessary closure creation and makes consistent with
preview completer.

kgryte committed:
refactor: make private properties read-only

kgryte committed:
fix: only perform preview completion when no collisions are possible

kgryte committed:
refactor: clean implementation with scoped variables

kgryte committed:
refactor: use helper functions

kgryte commented:
@Snehil-Shah As can be observed from the commit log, I did some fairly heavy refactoring. Notable changes:

1. Manual multi-line editing was removed. In general, PRs should have single concerns, and adding multi-line editing to the mix complicates review and implementation. Additionally, I think there is more to discuss in terms of whether using a modifier key to allow splitting expressions across multiple lines is the right approach. My preference is that we revisit this on https://github.com/stdlib-js/stdlib/issues/2060 and in a separate PR.
2. A public API for getting/setting settings was added to the REPL prototype. This allows programmatic control of settings from outside the REPL sandbox and also exposes REPL settings beyond the `settings` command.
3. Auto-matching logic was moved to a class, mirroring how we handle completion previews.
4. Reconciled auto-matching with completion previews. Completion preview logic had a bug when an identifier was in the middle of an expression, resulting in disjointed previews (e.g., `[func<|>]tion`, where `<|>` is the cursor.
5. The implementation was cleaned up to better match stdlib naming, variable, and comment conventions.

At this point, things sort of work; however, the implementation is not ideal. The principle problem is that the auto-matcher logic is not AST-aware, which makes for a frustrating UX. For example,

```
In [7]: ['a'<|>']
```

Typing `[`, `'`, `a`, `'`, results in an extra single quote, which is not what is wanted. Similarly,

```
In [7]: ['a']<|>]
```

results in an extra `]` after typing `]`. What is wanted is that the implementation "knows" that when I type `]`, I'm simply closing the array to complete the expression. Otherwise, to move beyond the closing `]`, I need to press the right arrow key which is not intuitive/instinctual.

In general, auto-matching brackets are there as hints to help guard against closing bracket omission. And as hints, they should appear and disappear (e.g., when I type over them). But because the implementation uses a naive, non-AST approach, it ends up getting more in the way than actually helping out.

Obviously, an AST approach is more involved and requires some R&D, so I think we have a choice: (1) we punt auto-closing to a future PR and simply merge the settings logic and APIs or (2) we go ahead and try to implement an AST-based approach.

Do you have a preference?

kgryte committed:
docs: add FIXME

Snehil-Shah commented:
> @Snehil-Shah As can be observed from the commit log, I did some fairly heavy refactoring. Notable changes:
> 
> 1. Manual multi-line editing was removed. In general, PRs should have single concerns, and adding multi-line editing to the mix complicates review and implementation. Additionally, I think there is more to discuss in terms of whether using a modifier key to allow splitting expressions across multiple lines is the right approach. My preference is that we revisit this on [[RFC]: add support for multi-line editing to the REPL #2060](https://github.com/stdlib-js/stdlib/issues/2060) and in a separate PR.
> 2. A public API for getting/setting settings was added to the REPL prototype. This allows programmatic control of settings from outside the REPL sandbox and also exposes REPL settings beyond the `settings` command.
> 3. Auto-matching logic was moved to a class, mirroring how we handle completion previews.
> 4. Reconciled auto-matching with completion previews. Completion preview logic had a bug when an identifier was in the middle of an expression, resulting in disjointed previews (e.g., `[func<|>]tion`, where `<|>` is the cursor.
> 5. The implementation was cleaned up to better match stdlib naming, variable, and comment conventions.
> 
> At this point, things sort of work; however, the implementation is not ideal. The principle problem is that the auto-matcher logic is not AST-aware, which makes for a frustrating UX. For example,
> 
> ```
> In [7]: ['a'<|>']
> ```
> 
> Typing `[`, `'`, `a`, `'`, results in an extra single quote, which is not what is wanted. Similarly,
> 
> ```
> In [7]: ['a']<|>]
> ```
> 
> results in an extra `]` after typing `]`. What is wanted is that the implementation "knows" that when I type `]`, I'm simply closing the array to complete the expression. Otherwise, to move beyond the closing `]`, I need to press the right arrow key which is not intuitive/instinctual.
> 
> In general, auto-matching brackets are there as hints to help guard against closing bracket omission. And as hints, they should appear and disappear (e.g., when I type over them). But because the implementation uses a naive, non-AST approach, it ends up getting more in the way than actually helping out.
> 
> Obviously, an AST approach is more involved and requires some R&D, so I think we have a choice: (1) we punt auto-closing to a future PR and simply merge the settings logic and APIs or (2) we go ahead and try to implement an AST-based approach.
> 
> Do you have a preference?

@kgryte you are right, we should be able to type over the auto-matched symbols. They should kind of be like preview completions except when not typed over, should automatically be added. I get the problem now. 
I think we can keep track of the auto-appended symbols somehow and allow typing over them. Can I get a small gist of the AST-based approach you have in mind?

kgryte committed:
refactor: add modules for specifying settings arguments amenable to auto-completion

kgryte committed:
refactor: add module exporting a regexp for matching settings APIs

kgryte committed:
refactor: add support for settings name auto-completion

kgryte commented:
@Snehil-Shah Looks like CodeMirror supports the behavior we are after. May be worthwhile studying its implementation: https://github.com/codemirror/autocomplete/blob/5ad2ebc861f2f61cdc943fc087a5bfb756a7d0fa/src/closebrackets.ts

kgryte committed:
refactor: defer to completer when receiving a TAB character

kgryte commented:
I have an idea; let me circle back.

kgryte committed:
refactor: use internal utility to support suppressing log messages

kgryte committed:
refactor: reorder cases in alphabetical order

kgryte committed:
refactor: fix order

kgryte committed:
refactor: update implementation to use AST

This commit ensures that opening symbols are not auto-closed in
strings and fixes unnecessary auto-insertion of closing characters
when closing characters are already present.

kgryte committed:
style: add lint warnings

kgryte committed:
style: address lint warning

kgryte committed:
test: add auto-match tests and move fixture

kgryte committed:
refactor: rename setting

kgryte committed:
docs: rename setting

kgryte committed:
refactor: remove unused property

kgryte committed:
fix: update property name

kgryte committed:
fix: allow auto-closing within template and object expressions

kgryte committed:
test: add RegExp test

kgryte committed:
fix: address completer bug when completing after special characters

kgryte committed:
refactor: add support for auto-deleting character pairs

kgryte committed:
refactor: move deletion logic to auto-closer due to shared state

kgryte committed:
fix: ensure corresponding closing symbol when auto-deleting pairs

kgryte committed:
docs: update copy

kgryte commented:
@Snehil-Shah Okay. I think this PR is ready for your review. I have reasonable confidence that things work, at least on the auto-closing side, where I've added a number of tests. Auto-deleting lacks tests, but I've done a fair amount of spot-checking in the REPL. Would you mind kicking the tires a bit and seeing if you uncover any bugs?

If everything seems okay, we can probably greenlight for merge. We are missing a number of tests, but perhaps those can be added in a follow-up PR?

Snehil-Shah committed:
docs: fix jsdoc

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
test: add tests for auto-deletion

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah commented:
@kgryte I tried it thoroughly, and it's working as expected.

Although another case where we can disable auto-closing quotations is just before/after a possible string literal like we have in IDEs. This is mostly the case when the user forgets to add quotations around strings and adds them later, and auto-closing becomes counter-intuitive.

Current:
```javascript
console.log('<|>'hello)
```
Expected:
```javascript
console.log('<|>hello)
```

I tried using acorn's findNodeAfter, but it would classify anything that starts with ' or " as a literal.
We can just use regex in this case to check the character just before/after the symbol like I did before
https://github.com/stdlib-js/stdlib/pull/1680/commits/db500417d709873e22f0c362c6e8e006db835dfe#diff-b3d2f6768907a0d9ef5adb03225c283c4fcfbc778f921a5af5bc1d2e0ff26f04 (from line 90)

Can I go ahead and implement this logic?
I also wrote some tests for the auto-deletion, please let me know if it needs changes

Another thing is do we need two separate settings for auto-deletion and auto-closing? Aren't both generally considered part of the same feature? Just a thought..

Snehil-Shah commented on lib/node_modules/@stdlib/repl/lib/main.js:
```diff
@@ -289,8 +301,14 @@ function REPL( options ) {
-		previewCompleter.beforeKeypress( data, key );
-		ttyWrite.call( self._rli, data, key );
+		var settings = self._settings;
+		if ( key && key.name === 'backspace' && settings.autoDeletePairs ) {
+			self._autoCloser.beforeKeypress( data, key );
+		}
```
@kgryte I think it would be better if we shift the `key.name === 'backspace'` check in the function itself as it's associated with autoCloser's logic and also limits the usage of the `beforeKeypress()` listener to just the case where the user enters backspace. What do you think?

kgryte replied:
@Snehil-Shah Yes, agreed. That seems reasonable. Feel free to go ahead and refactor.

Snehil-Shah committed:
Merge branch 'stdlib-js:develop' into repl

kgryte commented on lib/node_modules/@stdlib/repl/lib/auto_close_pairs.js:
```diff
@@ -65,8 +65,6 @@ function AutoCloser( rli ) {
 *
 * @name beforeKeypress
 * @memberof AutoCloser.prototype
-* @param {string} data - input data
```
@Snehil-Shah Why were these changes made? It is not uncommon for us to document parameters even if we don't use them in functions, as this allows for future dev reference. E.g., if we end up needing to modify `beforeKeypress` in the future to use `key`, we can reference the JSDoc, rather than track down the downstream callsite.

Snehil-Shah replied:
I apologize, I thought it was a mistake. Will revert.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -166,8 +166,6 @@ function moveBack( stream, N ) {
 * @private
 * @param {Object} t - test object
 * @param {Object} fixture - fixture object
-* @param {string} fixture.expression - incomplete expression string
```
@Snehil-Shah These should not have been removed. This is how we document required/expected properties. It is not enough that `fixture` is an object; in this case, `fixture` must have `expression` and `cursor` properties. We are not completely consistent on this, especially when an input object is expected to have many properties and/or properties/methods may frequently change, but, in this case, it is reasonable to document and leave in.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -232,8 +230,6 @@ function assertAutoClose( t, fixture, done ) {
 * @private
 * @param {Object} t - test object
 * @param {Object} fixture - fixture object
-* @param {string} fixture.expression - incomplete expression string
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/repl/lib/auto_close_pairs.js:
In general, the changes made in this commit should be reverted.

Snehil-Shah commented:
@kgryte I apologize for the JSDoc changes, I thought they were a mistake, will revert.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -284,6 +290,154 @@ function assertNoAutoClose( t, fixture, done ) {
+	moveBack( istream, N );
+
+	// Enable auto-close:
+	r.settings( 'autoClosePairs', true );
```
This isn't the setting you want to re-enable. You want to re-enable `autoDeletePairs`, as it is a separate setting. The `defaultSettings` function also needs to be updated to disable `autoDeletePairs` by default.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -284,6 +290,154 @@ function assertNoAutoClose( t, fixture, done ) {
+	r.settings( 'autoClosePairs', true );
+
+	// Delete the character in order to trigger auto-delete:
+	r._rli.write( null, {  // eslint-disable-line no-underscore-dangle
```
Did you try using an ANSI escape code written to `istream`? In general, we should avoid the assumption of private properties in tests and only test against public APIs.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -284,6 +290,154 @@ function assertNoAutoClose( t, fixture, done ) {
+	moveBack( istream, N );
+
+	// Enable auto-close:
+	r.settings( 'autoClosePairs', true );
+
+	// Delete the character in order to trigger auto-delete:
+	r._rli.write( null, {  // eslint-disable-line no-underscore-dangle
```
Same comments.

Snehil-Shah committed:
Revert "docs: fix jsdoc"

This reverts commit ff0a086587d9acade8dd92742cd544fb56c04cad.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
Given that `autoDeletePairs` is a separate setting, I am biased toward having the auto-delete tests in a separate test file. This would result in some duplication of test helper functions, but that is fine. As it is, by placing in the same file as `test.auto_close_pairs.js`, a developer only knows where to adjust auto-delete tests by actually reading test files. IMO it is preferable to be able to scan a directory and see the file with the corresponding settings name, so I can more readily know where to add/adjust tests.

kgryte commented:
@Snehil-Shah Thanks for adding auto-delete tests!

Snehil-Shah committed:
test: fix tests

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

kgryte commented:
> Another thing is do we need two separate settings for auto-deletion and auto-closing? Aren't both generally considered part of the same feature?

They are certainly related, but, while often they are implemented together, IMO they are still distinct. As a user, I may want auto-deletion, but not auto-closing because I may find auto-insertion annoying or want more fine-grained control as to when characters are inserted. And vice versa. For auto-deletion in particular, if I typed `[]` and then only want to delete the `[`, I need to resort to odd little workarounds (e.g., first inserting a space between the characters or deleting and then retyping `]`, etc). That may be enough for users to prefer one behavior but not the other.

kgryte commented:
> We can just use regex in this case to check the character just before/after the symbol like I did before.

Ah, okay, that was why you had that logic in there; I wasn't able to figure out from the comments alone what the intent was. But also, the regexp approach was too limited, as string content can be more than just ASCII. I think I have a solution for this and can push up a fix later today.

kgryte commented on lib/node_modules/@stdlib/repl/test/integration/test.auto_close_pairs.js:
```diff
@@ -342,13 +343,11 @@ function assertAutoDelete( t, fixture, done ) {
-	r._rli.write( null, {  // eslint-disable-line no-underscore-dangle
-		'name': 'backspace'
-	});
+	istream.write( '\x08' );
```
Prefer Unicode rather than hex.

Snehil-Shah committed:
refactor: separate auto-delete & auto-closing tests

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
refactor: move conditional to the method itself

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

kgryte commented on lib/node_modules/@stdlib/repl/lib/auto_close_pairs.js:
```diff
@@ -65,75 +65,77 @@ function AutoCloser( rli ) {
-		if ( ch !== line[ cursor ] ) {
-			debug( 'Opening symbol is not followed by a corresponding closing symbol. Skipping...' );
-			debug( 'Finished performing auto-delete.' );
+	if ( key && key.name === 'backspace' ) {
```
Rather than perform extra indentation like this; prefer early return. E.g.,

```
if ( !key || key.name !== 'backspace' ) {
    return false;
}
```

In general, we try and keep significant code blocks as left-aligned as possible, and, here, it better signifies intent. Namely, if the key is not a backspace, we don't need to process.

kgryte replied:
I'll go ahead and make this change, so as to avoid merge conflicts.

kgryte committed:
refactor: early return to avoid unnecessary indentation and rename parameter

kgryte committed:
docs: document properties

kgryte committed:
test: add separate test generation script

kgryte committed:
docs: update parameter type to reflect usage

kgryte committed:
refactor: move regular expression to separate file

kgryte committed:
refactor: access preceding character directly

kgryte committed:
refactor: replicate logic in private method

kgryte committed:
refactor: delegate open symbol processing to private method

kgryte committed:
refactor: copy logic for handling a closing symbol to a private method

kgryte committed:
refactor: delegate to private method for handling a closing symbol

kgryte committed:
refactor: implement handling of opening symbol during auto-deletion in private method

kgryte committed:
refactor: delegate to private method

kgryte committed:
docs: update param type

kgryte committed:
docs: fix copyright year

kgryte commented:
@Snehil-Shah Okay. I think this is in a pretty good state. You mind doing a final review to check that we didn't miss anything?

Snehil-Shah commented:
@kgryte Checked everything, looks good to go👍

kgryte commented:
@Snehil-Shah Thanks for checking! Let's go ahead and get this in. Any subsequent work can be addressed in follow-up PRs. Thanks for your work on this; this is really nice functionality to have.
